File: app/src/main/java/com/scop/org/minesweeper/generators/solver/Solver.java
package com.scop.org.minesweeper.generators.solver;

import com.scop.org.minesweeper.control.MainLogic;
import com.scop.org.minesweeper.elements.Tile;
import com.scop.org.minesweeper.utils.GridUtils;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.scop.org.minesweeper.elements.Tile.Status.*;

public abstract class Solver {
	private static MainLogic logic;
	protected static List<Tile> logicTiles;
	protected static Sketch sketch;

	public static void setSolver(MainLogic logic, Sketch sketch) {
		Solver.logic = logic;
		Solver.logicTiles = logic.getGrid().getGrid();
		Solver.sketch = sketch;
	}

	public abstract boolean analyze();

	public static void copyRevealed(){
		List<Tile> sketchTiles = sketch.getTiles();

		for (int i=0; i<logicTiles.size(); i++) {
			Tile tile = logicTiles.get(i);
			Tile sTile = sketchTiles.get(i);

			if (tile.getStatus() != COVERED && sTile.getStatus() == COVERED ) {
				int bombsNear = sTile.getBombsNear();
				sTile.setStatus(GridUtils.getTileStatus(bombsNear));

				if (bombsNear>0) sketch.addNumbered(i);
				else sketch.addUncovered(i);
			}
		}
	}

	protected void markBomb(int idx){
		if (logicTiles.get(idx).getStatus() == FLAG) return;
		logicTiles.get(idx).setStatus(FLAG);

		List<Tile> sketchTiles = sketch.getTiles();

		Tile tile = sketchTiles.get(idx);
		tile.setStatus(FLAG);
		tile.setCustomFlag('X');
		tile.defuseBomb();

		GridUtils.getNeighborsIdx(sketch.getGrid(), tile).stream()
				.peek(i-> {
					Tile t = sketchTiles.get(i);
					t.doesntHaveBombNear();
					if (t.isNumberVisible()) {
						int bombs = t.getBombsNear();
						t.setStatus(GridUtils.getTileStatus(bombs));
						if (bombs == 0) sketch.removeNumbered(i);
					}

				})
				.map(sketchTiles::get)
				.filter(t -> t.getBombsNear() == 0)
				.forEach(t -> t.setCustomFlag('0'));
	}

	protected void reveal(int idx){
		Tile t = logicTiles.get(idx);
		logic.reveal(t);

		tileUpdate(idx);
	}

	private void tileUpdate(int idx){
		Tile sTile = sketch.getTile(idx);
		if (sTile.isCovered()) {

			Tile tile = logicTiles.get(idx);

			if (tile.getStatus()==A0) {
				sketch.addUncovered(idx);
				sTile.setStatus(A0);
				for (int idxChild : GridUtils.getNeighborsIdx(sketch.getGrid(), sTile)){
					tileUpdate(idxChild);
				}

			} else {
				sketch.addNumbered(idx);
				sTile.setStatus(GridUtils.getTileStatus(sTile.getBombsNear()));
			}
		}
	}

	public static Set<Integer> getLeftBombs(){
		Set<Integer> li = new HashSet<>();

		for (int i : sketch.getNumbered()) {
			Tile sTile = sketch.getTiles().get(i);

			List<Integer> neigh = GridUtils.getNeighborsIdx(sketch.getGrid(), sTile);
			for (int n : neigh) {
				Tile t = logicTiles.get(n);
				if (t.getStatus() == COVERED && t.hasBomb())
					li.add(n);
			}
		}
		if (li.size() == 0) {
			for (int i = 0; i < logicTiles.size(); i++) {
				Tile t = logicTiles.get(i);
				if (t.getStatus() == COVERED && t.hasBomb())
					li.add(i);
			}
			if (li.size() > 0) {
				return null;
			}
		}
		return li;
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/solver/Solver.kt
package org.oar.minesweeper.generators.solver

import org.oar.minesweeper.control.MainLogic
import org.oar.minesweeper.elements.Tile
import org.oar.minesweeper.elements.TileStatus
import org.oar.minesweeper.utils.GridUtils.getNeighborsIdx

abstract class Solver {

    companion object {
        private lateinit var logic: MainLogic
        lateinit var logicTiles: List<Tile>
        lateinit var sketch: Sketch

        fun setSolver(logic: MainLogic, sketch: Sketch) {
            Companion.logic = logic
            logicTiles = logic.grid.tiles
            Companion.sketch = sketch
        }

        fun copyRevealed() {
            val sketchTiles = sketch.tiles
            for (i in logicTiles.indices) {
                val tile = logicTiles[i]
                val sTile = sketchTiles[i]
                if (tile.status !== TileStatus.COVERED && sTile.status === TileStatus.COVERED) {
                    val bombsNear = sTile.bombsNear
                    sTile.status = TileStatus.findByTileNumber(bombsNear)
                    if (bombsNear > 0) sketch.addNumbered(i) else sketch.addUncovered(i)
                }
            }
        }

        val leftBombs: Set<Int>?
            get() {
                val li: MutableSet<Int> = HashSet()
                for (i in sketch.numbered) {
                    val sTile = sketch.tiles[i]
                    val neigh = sketch.grid.getNeighborsIdx(sTile)
                    for (n in neigh) {
                        val t = logicTiles[n]
                        if (t.status === TileStatus.COVERED && t.hasBomb) li.add(n)
                    }
                }
                if (li.size == 0) {
                    for (i in logicTiles.indices) {
                        val t = logicTiles[i]
                        if (t.status === TileStatus.COVERED && t.hasBomb) li.add(i)
                    }
                    if (li.size > 0) {
                        return null
                    }
                }
                return li
            }
    }


    abstract fun analyze(): Boolean

    protected fun markBomb(index: Int) {
        if (logicTiles[index].status === TileStatus.FLAG) return
        logicTiles[index].status = TileStatus.FLAG

        val sketchTiles = sketch.tiles

        val tile = sketchTiles[index]
        tile.status = TileStatus.FLAG
        tile.customFlag.add('X')
        tile.hasBomb = false

        sketch.grid.getNeighborsIdx(tile)
            .onEach { idx ->
                val t = sketchTiles[idx]
                t.doesntHaveBombNear()
                if (t.isNumberVisible) {
                    val bombs = t.bombsNear
                    t.status = TileStatus.findByTileNumber(bombs)
                    if (bombs == 0) sketch.removeNumbered(idx)
                }
            }
            .map { idx -> sketchTiles[idx] }
            .filter { it.bombsNear == 0 }
            .forEach { it.customFlag.add('0') }
    }

    protected fun reveal(idx: Int) {
        val t = logicTiles[idx]
        logic.reveal(t)
        tileUpdate(idx)
    }

    private fun tileUpdate(idx: Int) {
        val sTile = sketch.getTile(idx)
        if (sTile.isCovered) {
            val tile = logicTiles[idx]

            if (tile.status === TileStatus.A0) {
                sketch.addUncovered(idx)
                sTile.status = TileStatus.A0

                sketch.grid.getNeighborsIdx(sTile)
                    .forEach { idxChild -> tileUpdate(idxChild) }

            } else {
                sketch.addNumbered(idx)
                sTile.status = TileStatus.findByTileNumber(sTile.bombsNear)
            }
        }
    }
}